drivers/pwm/pwm_nrfx.c

```diff
@@ -5,93 +5,236 @@
  */
 #include <nrfx_pwm.h>
 #include <pwm.h>
+#include <hal/nrf_gpio.h>
+#include <stdbool.h>
 
 #define LOG_LEVEL CONFIG_PWM_LOG_LEVEL
 #include <logging/log.h>
 LOG_MODULE_REGISTER(pwm_nrfx);
 
-#define PWM_NRFX_CH_VALUE_NORMAL        (1UL << 15)
+#define PWM_NRFX_CH_POLARITY_MASK       BIT(15)
+#define PWM_NRFX_CH_PULSE_CYCLES_MASK   BIT_MASK(15)
+#define PWM_NRFX_CH_VALUE_NORMAL        PWM_NRFX_CH_POLARITY_MASK
 #define PWM_NRFX_CH_VALUE_INVERTED      (0)
+#define PWM_NRFX_CH_PIN_MASK            ~NRFX_PWM_PIN_INVERTED
 
 struct pwm_nrfx_config {
 	nrfx_pwm_t pwm;
-	nrfx_pwm_config_t config;
+	nrfx_pwm_config_t initial_config;
 	nrf_pwm_sequence_t seq;
 };
 
 struct pwm_nrfx_data {
+	u32_t period_cycles;
 	u16_t current[NRF_PWM_CHANNEL_COUNT];
-	u16_t top_value;
+	u16_t countertop;
+	u8_t  prescaler;
 };
 
+static u32_t pwm_period_check_and_set(const struct pwm_nrfx_config *pconfig,
+				      struct pwm_nrfx_data *pdata,
+				      u32_t pwm,
+				      u32_t period_cycles)
+{
+	NRF_PWM_Type *pwm_instance = pconfig->pwm.p_registers;
+
+	if (!nrfx_pwm_is_stopped(&pconfig->pwm)) {
+		/* Succeed if requested period matches already used period */
+		if (period_cycles == pdata->period_cycles) {
+			return 0;
+		}
+
+		/* Fail if requested period != already running period */
+		LOG_ERR("Fail:requested period cycles:%d, != used %d\n",
+			period_cycles, pdata->period_cycles);
+		return -EINVAL;
+	}
+
+	/* Check if period_cycles is above COUNTERTOP MAX value, if so, we
+	 * have to see if we can change frequency to something that will fit
+	 */
+	if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
+		/* See if there is a prescaler that will make it work: */
+		bool matching_prescaler_found = false;
+
+		/* Go through all available prescaler values on device.
+		 * nRF52832 has 0-7 (Div1 - Div128)
+		 */
+		for (u8_t prescaler = 0;
+		     prescaler <= PWM_PRESCALER_PRESCALER_Msk;
+		     prescaler++) {
+			u32_t new_countertop = period_cycles >> prescaler;
+
+			/* If we find value that fits, set it, continue */
+			if (new_countertop <= PWM_COUNTERTOP_COUNTERTOP_Msk) {
+				pdata->prescaler = prescaler;
+				pdata->countertop = new_countertop;
+				pdata->period_cycles = period_cycles;
+				matching_prescaler_found = true;
+				break;
+			}
+		}
+
+		/* Check if able to find matching prescaler and countertop */
+		if (matching_prescaler_found == false) {
+			LOG_ERR("Prescaler for period_cycles %d not found.\n",
+				period_cycles);
+			return -EINVAL;
+		}
+	} else {
+		/* If period_cycles fit with standard prescaler,
+		 * set it directly
+		 */
+		pdata->prescaler = 1;
+		pdata->countertop = period_cycles;
+		pdata->period_cycles = period_cycles;
+	}
+
+	/* Write new PRESCALER and COUNTERTOP to PWM instance */
+	nrf_pwm_configure(pwm_instance,
+		pdata->prescaler,
+		pconfig->initial_config.count_mode,
+		pdata->countertop);
```

Comment:
Please align these lines with first argument to function.

Reply:
done. As mentioned earlier, also removed p prefix on variable names.